File: python/2022/day12.py
# vi: set shiftwidth=4 tabstop=4 expandtab:
import datetime
import os
import collections


resource_dir = os.path.dirname(os.path.abspath(__file__)) + "/../../resources/"

def get_grid_from_lines(string):
    start = None
    dest = None
    grid = {}
    for i, l in enumerate(string.splitlines()):
        for j, val in enumerate(l):
            pos = (i, j)
            if val == "S":
                start = pos
                val = "a"
            elif val == "E":
                dest = pos
                val = "z"
            grid[pos] = ord(val) - ord("a")
    assert start is not None
    assert dest is not None
    return grid, start, dest


def get_grid_from_file(file_path=resource_dir + "year2022_day12_input.txt"):
    with open(file_path) as f:
        return get_grid_from_lines(f.read())


def get_neighbours(x, y):
    return ((x + dx, y + dy) for (dx, dy) in ((-1, 0), (+1, 0), (0, -1), (0, +1)))


def get_accessible_neighbours(grid, uphill):
    access = {}
    for pos, val in grid.items():
        d = access.setdefault(pos, {})
        for pos2 in get_neighbours(*pos):
            val2 = grid.get(pos2)
            if val2 is not None and (
                (val2 <= val + 1) if uphill else (val2 + 1 >= val)
            ):
                d[pos2] = 1
    return access


def get_distances(grid, start, uphill):
    distances = {}
    queue = collections.deque([(0, start)])
    neigh = get_accessible_neighbours(grid, uphill)
    while queue:
        d, pos = queue.popleft()
        if pos not in distances or distances[pos] > d:
            distances[pos] = d
            queue.extend((d + d2, pos2) for pos2, d2 in neigh[pos].items())
    return distances


def get_path(grid, start, dest):
    return get_distances(grid, start, uphill=True)[dest]


def get_path2(grid, dest):
    return min(
        d
        for pos, d in get_distances(grid, dest, uphill=False).items()
        if grid[pos] == 0
    )


def run_tests():
    grid, start, dest = get_grid_from_lines(
        """Sabqponm
abcryxxl
accszExk
acctuvwj
abdefghi"""
    )
    assert get_path(grid, start, dest) == 31
    assert get_path2(grid, dest) == 29


def get_solutions():
    grid, start, dest = get_grid_from_file()
    print(get_path(grid, start, dest) == 423)
    print(get_path2(grid, dest) == 416)


if __name__ == "__main__":
    begin = datetime.datetime.now()
    run_tests()
    get_solutions()
    end = datetime.datetime.now()
    print(end - begin)

File: rust/src/2022/day12.rs
// vi: set shiftwidth=4 tabstop=4 expandtab:
use common::input::check_answer;
use common::input::get_answers;
use common::input::get_file_content;
use common::point_module;
use std::collections::HashMap;
use std::collections::VecDeque;
use std::time::Instant;

const INPUT_FILEPATH: &str = "../resources/year2022_day12_input.txt";
const ANSWERS_FILEPATH: &str = "../resources/year2022_day12_answer.txt";

type Int = u32;
type Point = point_module::Point<isize>;
type Grid = HashMap<Point, Int>;
type InputContent = (Grid, Point, Point);

#[allow(clippy::cast_possible_wrap)]
fn get_input_from_str(string: &str) -> InputContent {
    let mut start = None;
    let mut end = None;
    let mut grid = Grid::new();
    for (i, l) in string.lines().enumerate() {
        for (j, val) in l.chars().enumerate() {
            let pos = Point {
                x: i as isize,
                y: j as isize,
            };
            let mut c = val;
            if val == 'S' {
                start = Some(pos);
                c = 'a';
            } else if val == 'E' {
                end = Some(pos);
                c = 'z';
            }
            grid.insert(pos, c as Int - 'a' as Int);
        }
    }
    (grid, start.unwrap(), end.unwrap())
}

fn get_neighbours(Point { x, y }: &Point) -> Vec<Point> {
    [(-1, 0), (1, 0), (0, -1), (0, 1)]
        .iter()
        .map(|(dx, dy)| Point {
            x: x + dx,
            y: y + dy,
        })
        .collect()
}

fn get_accessible_neighbours(grid: &Grid, uphill: bool) -> HashMap<Point, HashMap<Point, Int>> {
    let mut access = HashMap::new();
    for (pos, &val) in grid {
        let d = access.entry(*pos).or_insert_with(HashMap::new);
        for pos2 in get_neighbours(pos) {
            if let Some(&val2) = grid.get(&pos2) {
                if if uphill {
                    val2 <= val + 1
                } else {
                    val <= val2 + 1
                } {
                    d.insert(pos2, 1);
                }
            }
        }
    }
    access
}

fn get_distances(grid: &Grid, start: Point, uphill: bool) -> HashMap<Point, Int> {
    let mut distances = HashMap::new();
    let mut deq = VecDeque::from([(0, start)]);
    let neigh = get_accessible_neighbours(grid, uphill);
    while let Some((d, pos)) = deq.pop_front() {
        if let Some(&old_d) = distances.get(&pos) {
            if old_d <= d {
                continue;
            }
        }
        distances.insert(pos, d);
        let neigh2 = &neigh[&pos];
        for (pos2, d2) in neigh2 {
            deq.push_back((d + d2, *pos2));
        }
    }
    distances
}

fn part1((grid, start, dest): &InputContent) -> Int {
    get_distances(grid, *start, true)[dest]
}

fn part2((grid, _start, dest): &InputContent) -> Int {
    *get_distances(grid, *dest, false)
        .iter()
        .filter(|(pos, _)| grid[pos] == 0)
        .map(|(_, d)| d)
        .min()
        .unwrap()
}

fn main() {
    let before = Instant::now();
    let data = get_input_from_str(&get_file_content(INPUT_FILEPATH));
    let (ans, ans2) = get_answers(ANSWERS_FILEPATH);
    let solved = true;
    let res = part1(&data);
    check_answer(&res.to_string(), ans, solved);
    let res2 = part2(&data);
    check_answer(&res2.to_string(), ans2, solved);
    println!("Elapsed time: {:.2?}", before.elapsed());
}

#[cfg(test)]
mod tests {
    use super::*;

    const EXAMPLE: &str = "Sabqponm
abcryxxl
accszExk
acctuvwj
abdefghi";

    #[test]
    fn test_part1() {
        assert_eq!(part1(&get_input_from_str(EXAMPLE)), 31);
    }

    #[test]
    fn test_part2() {
        assert_eq!(part2(&get_input_from_str(EXAMPLE)), 29);
    }
}
